### PR TITLE
add 1.4.13.4 for SZ

### PR DESCRIPTION
**Pull request by disheng222**
We release a new version for SZ, in which there are some useful functions added.
Details about versions are README of http://github.com/disheng222/SZ
Thanks.

### CODE REVIEW

File: var/spack/repos/builtin/packages/sz/package.py [hunk 1]
```diff
@@ -30,10 +30,13 @@ class Sz(AutotoolsPackage):
     """Error-bounded Lossy Compressor for HPC Data."""
 
     homepage = "https://collab.cels.anl.gov/display/ESR/SZ"
-    url      = "https://github.com/disheng222/SZ/archive/v1.4.11.0.tar.gz"
+    url      = "https://github.com/disheng222/SZ/archive/v1.4.13.4.zip"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/sz/package.py]: If you change the extension, none of the older versions will download. Can you keep it `.tar.gz` and update the checksums?

File: var/spack/repos/builtin/packages/sz/package.py [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
-# Copyright (c) 2013-2018, Lawrence Livermore National Security, LLC.
+# Copyright (c) 2013-2017, Lawrence Livermore National Security, LLC.
 # Produced at the Lawrence Livermore National Laboratory.
 #
 # This file is part of Spack.
 # Created by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
 # LLNL-CODE-647188
 #
-# For details, see https://github.com/spack/spack
+# For details, see https://github.com/llnl/spack
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/sz/package.py]: Can you undo the changes to the license? `github.com/llnl/spack` is out of date.

**Comment by adamjstewart** [comment]
Is this a duplicate of #8372?

**Comment by disheng222** [comment]
Hi, 
I changed the package.py according to your comment. 
Please feel free to let me know more questions/suggestions if any.
Thanks.

**Comment by disheng222** [comment]
Hi Adam,
I just changed the package.py and submitted a new "pull request" based on
your comment.

Please feel free to let me know more questions/suggestions if any.
I submitted a pull request yesterday, but I guess that one was rejected.

Thanks.

Best,
Sheng

2018-06-05 13:41 GMT-05:00 Adam J. Stewart <notifications@github.com>:

**Comment by adamjstewart** [comment]
> I submitted a pull request yesterday, but I guess that one was rejected.

It wasn't rejected, it's still open.

File: var/spack/repos/builtin/packages/sz/package.py [hunk 3]
```diff
@@ -25,15 +25,15 @@
 
 from spack import *
 
-
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/sz/package.py]: Deleting this line will cause flake8 to fail.

**Comment by ax3l** [comment]
@disheng222 when you are adding new `sz` releases, can you please verify the `adios` package still builds? `sz` is constantly changing it's API in patch-level releases (e.g. `1.4.12` to `1.4.13`) which breaks the adios build, e.g. 1.13.1 which is in spack :-)

no problem, but please consider running `spack install adios` after you added a new release.

**Comment by ax3l** [comment]
Fixed it in  #8408

Also: @disheng222 congrats on the new SZ release! :)

**Comment by disheng222** [comment]
Hi Axel,
Thanks for the information.
We changed SZ API only once in history and I notified Adios developer about
this. They said they will tune it in their release.
I will contact Adios developer for further information.
Thanks!

Best,
Sheng

**Comment by ax3l** [comment]
@disheng222 thanks for the information!
Huh, confusing - I got compile errors for adios on changes to `1.4.10`, `.11` and `.12`. Any idea what could be causing this?

Thanks,
Axel

**Comment by disheng222** [comment]
Hi Jon,
Axel said there are compile errors for Adios with SZ, probably because of
changed API from 1.4.9 to 1.4.10.
I checked the API. I guess you are using SZ_compress_args function, and it
did have some changes from 1.4.9 and 1.4.10.
1.4.9:
unsigned char *SZ_compress_args(int dataType, void *data, int *outSize, int
errBoundMode, double absErrBound, double relBoundRatio, int r5, int r4, int
r3, int r2, int r1);
1.4.10:
unsigned char* SZ_compress_args(int dataType, void *data, size_t *outSize,
int errBoundMode, double absErrBound,
double relBoundRatio, double pwrBoundRatio, int pwrType, size_t r5, size_t
r4, size_t r3, size_t r2, size_t r1);

In the 1.4.10+, pwrBoundRatio means point-wise relative error bound ratio
and pwrType could be set to 0 for simplicity.
If the users set the error bound mode to be absolute error bound or
value-range based relative error bound, pwrBoundRatio and pwrType will be
ignored.

Could you change the corresponding codes in ADIOS to fix the compile issue?
Thanks in advance!

Best,
Sheng

**Comment by disheng222** [comment]
Hi Jong,
I just changed the API again in order to make it easy for you.
I removed the "pwrType" argument from SZ_compress_args(), because this
argument will be deprecated in the future SZ version.
If you used SZ_compress_args() in your code, please change it
correspondingly, as follows. Thanks.
unsigned char* SZ_compress_args(int dataType, void *data, size_t *outSize,
int errBoundMode, double absErrBound,
double relBoundRatio, double pwrBoundRatio, size_t r5, size_t r4, size_t
r3, size_t r2, size_t r1)

If you actually didn't use SZ_compress_args() but SZ_compress() to do the
compression, you don't need to change your code.
Any questions are welcome!

Best,
Sheng

**Comment by ax3l** [comment]
@disheng222 @jychoi-hpc the problem I am currently seeing is that ADIOS 1.13.1 does not compile with SZ 1.4.13.* but only with 1.4.12.*

ADIOS 1.13.0 only works with SZ 1.4.11 (not any later) and ADIOS 1.12.0 only works with SZ 1.4.10 (not any later). Maybe you can figure out the source of this? According to @disheng222 the public API has not changed between the releases.

**Comment by jychoi-hpc** [comment]
Just did a quick look-through, ADIOS is not using SZ_compress_args. Instead, it calls the following SZ APIs:
```
SZ_Init
SZ_Init_Params(&sz);
SZ_compress
SZ_decompress
SZ_Finalize
```
One time, I remember there was a change of size from int to size_t one of the parameters. I guess some compilers will make only warning or you will get a hard error.

Regarding the using the most recent version of SZ, I have just checked using ADIOS 1.13.1 with SZ 1.4.13.5 and found that there is some changes in sz_params structure. Basically, a few members are removed (dataEndianType, sysEndianType, etc). I can fix this from the ADIOS side.

**Comment by ax3l** [comment]
@disheng222 are the APIs @jychoi-hpc mentions stable or should other SZ routines be used?

**Comment by disheng222** [comment]
The SZ routines are stable. Based on Jong's answers, I guess the reason is
that I deleted some deprecated parameters in a global data structure
instead of API.
I believe Jong already knows how to fix it in Adios.
Best,
Sheng

**Comment by ax3l** [comment]
Alright, thanks @disheng222!
